keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1290 @@
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel`: `06/14/2023` - `83.2`
+- `k8s.io/apiextensions-apiserver/pkg/apiserver/schema/listtype`: `06/14/2023` - `95.8`
+- `k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta`: `06/14/2023` - `83.9`
+- `k8s.io/apiextensions-apiserver/pkg/registry/customresource`: `06/14/2023` - `57.5`
+- `k8s.io/apiextensions-apiserver/pkg/apiserver/validation`: `06/14/2023` - `85.8`
+- `k8s.io/kube-openapi/pkg/validation/validate` : `06/14/2023` - `97.2`
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+1. Show normal creation path is unaffected
+2. Tests that ratchet different classes of validations:
+    - JSON-Schema fields (minItems, dependencies, etc.)
+    - CRD-Validation Rules
+    - ListSets/Maps
+3. Tests that certain classes of valdiations are not ratcheted:
+    - Nested Validations (not, oneOf, allOf, anyOf)
+    - Transition Rules
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+It would be nice to have a e2e test that tests ratcheting over a k8s upgrade
+which tightens a validation.
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+
+#### Alpha
+- Feature Flag added
+- Integration Tests implemented and enabled
+- Incomplete-but-correct subset of ratcheting rules implemented under flag
+
+#### Beta
+- Gather feedback from developers
+- Additional Testing added as needed
+- CEL Validation Rules Ratcheting Implemented
+- Incomplete-but-correct subset of ratcheting rules implemented under flag
```

Comment:
why would this still be incomplete in beta?

Reply:
"incomplete" in the semantic/mathematical sense. We may have certain properties we cannot ratchet or can only ratchet under some conditions

Reply:
Changed this section to specifically call our the parts of object ratcheting to be implemented during beta and removed incompleteness terminology.

Reply:
I would even claim "there is no everything", or at a certain point it becomes highly subjective which ratcheting is what to expect and which is not. Hence, we will move those cases out of the feature gate that we are comfortable with.